File: packages/react-native-enriched-markdown/ios/utils/StreamingMarkdownFilter.m
#import "StreamingMarkdownFilter.h"

static BOOL ENRMLineIsBlank(NSString *line)
{
  return [line stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]].length == 0;
}

static BOOL ENRMLineIsBlockMathDelimiter(NSString *line)
{
  return [[line stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]] isEqualToString:@"$$"];
}

static BOOL ENRMLineLooksLikeTableRow(NSString *line)
{
  NSString *trimmed = [line stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
  return [trimmed hasPrefix:@"|"];
}

static NSUInteger ENRMPipeCount(NSString *line)
{
  NSUInteger count = 0;
  for (NSUInteger i = 0; i < line.length; i++) {
    if ([line characterAtIndex:i] == '|') {
      count++;
    }
  }
  return count;
}

static BOOL ENRMLineLooksLikeTableSeparator(NSString *line)
{
  NSString *trimmed = [line stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
  if (trimmed.length == 0) {
    return NO;
  }
  if ([trimmed characterAtIndex:0] != '|') {
    return NO;
  }
  BOOL hasTripleDash = NO;
  NSUInteger dashRun = 0;
  for (NSUInteger i = 0; i < trimmed.length; i++) {
    unichar ch = [trimmed characterAtIndex:i];
    if (ch == '-') {
      dashRun++;
      if (dashRun >= 3) {
        hasTripleDash = YES;
      }
    } else {
      dashRun = 0;
      if (ch != '|' && ch != ':' && ch != ' ') {
        return NO;
      }
    }
  }
  return hasTripleDash;
}

static NSUInteger *ENRMBuildLineOffsets(NSArray<NSString *> *lines, NSUInteger count)
{
  NSUInteger *offsets = (NSUInteger *)calloc(count, sizeof(NSUInteger));
  NSUInteger currentOffset = 0;
  for (NSUInteger i = 0; i < count; i++) {
    offsets[i] = currentOffset;
    currentOffset += lines[i].length + 1;
  }
  return offsets;
}

static NSString *ENRMRemovePendingStreamingMathBlock(NSString *markdown, NSArray<NSString *> *lines)
{
  NSInteger lastUnclosedDelimiterIndex = -1;

  for (NSUInteger i = 0; i < lines.count; i++) {
    if (ENRMLineIsBlockMathDelimiter(lines[i])) {
      lastUnclosedDelimiterIndex = lastUnclosedDelimiterIndex == -1 ? (NSInteger)i : -1;
    }
  }

  if (lastUnclosedDelimiterIndex == -1) {
    return markdown;
  }

  NSUInteger *offsets = ENRMBuildLineOffsets(lines, lines.count);
  NSString *result = [markdown substringToIndex:offsets[(NSUInteger)lastUnclosedDelimiterIndex]];
  free(offsets);
  return result;
}

static NSString *ENRMRemovePendingStreamingTableBlock(NSString *markdown, NSArray<NSString *> *lines,
                                                      ENRMTableStreamingMode tableMode)
{
  NSInteger lastNonBlankLineIndex = -1;

  for (NSInteger i = (NSInteger)lines.count - 1; i >= 0; i--) {
    if (!ENRMLineIsBlank(lines[(NSUInteger)i])) {
      lastNonBlankLineIndex = i;
      break;
    }
  }

  if (lastNonBlankLineIndex == -1) {
    return markdown;
  }

  if ((NSUInteger)lastNonBlankLineIndex + 1 < lines.count - 1) {
    return markdown;
  }

  NSInteger blockStartIndex = lastNonBlankLineIndex;
  while (blockStartIndex > 0 && !ENRMLineIsBlank(lines[(NSUInteger)blockStartIndex - 1])) {
    blockStartIndex--;
  }

  BOOL blockLooksLikeTable = NO;
  for (NSInteger i = blockStartIndex; i <= lastNonBlankLineIndex; i++) {
    NSString *line = lines[(NSUInteger)i];
    if (!ENRMLineLooksLikeTableRow(line)) {
      return markdown;
    }
    blockLooksLikeTable = YES;
  }

  if (!blockLooksLikeTable) {
    return markdown;
  }

  NSUInteger *offsets = ENRMBuildLineOffsets(lines, lines.count);

  if (tableMode == ENRMTableStreamingModeProgressive) {
    NSInteger tableLineCount = lastNonBlankLineIndex - blockStartIndex + 1;

    if (tableLineCount < 2 || !ENRMLineLooksLikeTableSeparator(lines[(NSUInteger)blockStartIndex + 1])) {
      NSString *result = [markdown substringToIndex:offsets[(NSUInteger)blockStartIndex]];
      free(offsets);
      return result;
    }

    if (tableLineCount > 2) {
      NSString *lastRow = lines[(NSUInteger)lastNonBlankLineIndex];
      NSString *lastRowTrimmed = [lastRow stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
      NSString *headerRow = lines[(NSUInteger)blockStartIndex];
      if (![lastRowTrimmed hasSuffix:@"|"] || ENRMPipeCount(lastRow) < ENRMPipeCount(headerRow)) {
        NSString *result = [markdown substringToIndex:offsets[(NSUInteger)lastNonBlankLineIndex]];
        free(offsets);
        return result;
      }
    }

    free(offsets);
    return markdown;
  }

  NSString *result = [markdown substringToIndex:offsets[(NSUInteger)blockStartIndex]];
  free(offsets);
  return result;
}

NSString *ENRMRenderableMarkdownForStreaming(NSString *markdown, ENRMTableStreamingMode tableMode)
{
  NSArray<NSString *> *lines = [markdown componentsSeparatedByString:@"\n"];
  NSString *afterMath = ENRMRemovePendingStreamingMathBlock(markdown, lines);
  NSArray<NSString *> *linesForTable =
      (afterMath.length == markdown.length) ? lines : [afterMath componentsSeparatedByString:@"\n"];
  return ENRMRemovePendingStreamingTableBlock(afterMath, linesForTable, tableMode);
}

File: packages/react-native-enriched-markdown/android/src/main/java/com/swmansion/enriched/markdown/utils/common/StreamingMarkdownFilter.kt
package com.swmansion.enriched.markdown.utils.common

enum class TableStreamingMode {
  HIDDEN,
  PROGRESSIVE,
}

/**
 * Pre-parse filter that hides incomplete trailing tables and block math
 * during streaming. A table is considered complete only after a blank
 * separator line follows it; a block math (`$$`) is complete only when
 * a closing `$$` exists.
 */
object StreamingMarkdownFilter {
  fun renderableMarkdownForStreaming(
    markdown: String,
    tableMode: TableStreamingMode = TableStreamingMode.PROGRESSIVE,
  ): String {
    val lines = markdown.split("\n")
    val afterMath = removePendingStreamingMathBlock(markdown, lines)
    val linesForTable = if (afterMath.length == markdown.length) lines else afterMath.split("\n")
    return removePendingStreamingTableBlock(afterMath, linesForTable, tableMode)
  }

  private fun removePendingStreamingMathBlock(
    markdown: String,
    lines: List<String>,
  ): String {
    var lastUnclosedDelimiterIndex = -1

    for (i in lines.indices) {
      if (lineIsBlockMathDelimiter(lines[i])) {
        lastUnclosedDelimiterIndex = if (lastUnclosedDelimiterIndex == -1) i else -1
      }
    }

    if (lastUnclosedDelimiterIndex == -1) return markdown

    val offsets = buildLineOffsets(lines)
    return markdown.substring(0, offsets[lastUnclosedDelimiterIndex])
  }

  private fun removePendingStreamingTableBlock(
    markdown: String,
    lines: List<String>,
    tableMode: TableStreamingMode,
  ): String {
    var lastNonBlankLineIndex = -1

    for (i in lines.indices.reversed()) {
      if (!lineIsBlank(lines[i])) {
        lastNonBlankLineIndex = i
        break
      }
    }

    if (lastNonBlankLineIndex == -1) return markdown

    if (lastNonBlankLineIndex + 1 < lines.size - 1) return markdown

    var blockStartIndex = lastNonBlankLineIndex
    while (blockStartIndex > 0 && !lineIsBlank(lines[blockStartIndex - 1])) {
      blockStartIndex--
    }

    var blockLooksLikeTable = false
    for (i in blockStartIndex..lastNonBlankLineIndex) {
      if (!lineLooksLikeTableRow(lines[i])) return markdown
      blockLooksLikeTable = true
    }

    if (!blockLooksLikeTable) return markdown

    val offsets = buildLineOffsets(lines)

    if (tableMode == TableStreamingMode.PROGRESSIVE) {
      val tableLineCount = lastNonBlankLineIndex - blockStartIndex + 1

      if (tableLineCount < 2 || !lineLooksLikeTableSeparator(lines[blockStartIndex + 1])) {
        return markdown.substring(0, offsets[blockStartIndex])
      }

      if (tableLineCount > 2) {
        val lastRow = lines[lastNonBlankLineIndex]
        val lastRowTrimmed = lastRow.trim()
        val headerRow = lines[blockStartIndex]
        if (!lastRowTrimmed.endsWith("|") || pipeCount(lastRow) < pipeCount(headerRow)) {
          return markdown.substring(0, offsets[lastNonBlankLineIndex])
        }
      }

      return markdown
    }

    return markdown.substring(0, offsets[blockStartIndex])
  }

  private fun lineIsBlank(line: String): Boolean = line.isBlank()

  private fun lineIsBlockMathDelimiter(line: String): Boolean = line.trim() == "$$"

  private fun lineLooksLikeTableRow(line: String): Boolean {
    val trimmed = line.trim()
    return trimmed.startsWith("|")
  }

  private fun lineLooksLikeTableSeparator(line: String): Boolean {
    val trimmed = line.trim()
    if (trimmed.isEmpty()) return false
    if (trimmed[0] != '|') return false
    var hasTripleDash = false
    var dashRun = 0
    for (ch in trimmed) {
      if (ch == '-') {
        dashRun++
        if (dashRun >= 3) hasTripleDash = true
      } else {
        dashRun = 0
        if (ch != '|' && ch != ':' && ch != ' ') return false
      }
    }
    return hasTripleDash
  }

  private fun pipeCount(line: String): Int {
    var count = 0
    for (ch in line) {
      if (ch == '|') count++
    }
    return count
  }

  private fun buildLineOffsets(lines: List<String>): IntArray {
    val offsets = IntArray(lines.size)
    var currentOffset = 0
    for (i in lines.indices) {
      offsets[i] = currentOffset
      currentOffset += lines[i].length + 1
    }
    return offsets
  }
}
